### ISSUE TITLE
Link that opens PDF at specific page

### ISSUE THREAD

**Issue by sanjarcode** [zim-desktop-wiki/zim-desktop-wiki#1386]
**Goal**
Add link which opens a PDF at a specific page/bookmark.

**Context**
I have a lot of PDFs attached to my Zim notebook. I wish to create links in my notes where I can reference specific pages/bookmarks in the PDF.

**Problem**
This is possible for PDFs stored online because I can set the link to
```https://...some_website.../my_doc.pdf#page=23``` which works. The PDF opens in the browser, which is OK since it's only for peeking at a single page.

For local ones ```file:///home/superman/my_doc.pdf#page=23``` works fine in the browser, but Zim automatically edits the link's prefix, for example, `file:///home/superman` changes to `~`. On clicking the link, the presence of `~` makes Zim open the link as a local file using `xdg-open`, which obviously has no clue about the `#page` convention and treats it as part of the file name. Consequently, Zim displays a 'No such file or folder' error.

**Solutions**
1. Add option to disable truncating ```file:///``` in link hrefs, which is otherwise very important in Zim because it handles many things.
2. Ability to run a small snippet of bash-like code(like a link), for example, triggering ```evince ~/my_doc.pdf -p 23``` by a single click would be great. An advantage is that this solution does not mess with the link behavior at all, so it's harmless and can be implemented as a plugin. Plus it can be used to do almost anything from within a notebook, like starting up a server or project, making the notebook somewhat interactive.

Suggestions are welcome

**Comment by pfalcon** [comment]
> On clicking the link, the presence of ~ makes Zim open the link as a local file using xdg-open

> Ability to run a small snippet of bash-like code(like a link),

You have that ability. Update your system mime types to run a script, which will parse `#page=` at the end of "filename", and run `evince ~/my_doc.pdf -p 23` or whatever.

**Comment by sojusnik** [comment]
@pfalcon Can you graciously describe this process step-by-step?

**Comment by pfalcon** [comment]
To do that, I would need to read the docs and use google, and that would wait in my current TODO queue for years (given that I don't need that feature). But why would you want to be blocked on me for years, instead of reading docs and googling up yourself?

**Comment by roland-ruedenauer** [comment]
Maybe you missed this:
If you right-click on a pdf document attached to a page you should see an entry offering to copy a link to the document.
Then you can go to any page in your wiki and insert the link (also edit the link text as usual).
The link created should look like `[[../Task_3/Task_3-1/Test.pdf]]` in the page source.

**Comment by sanjarcode** [comment]
> Maybe you missed this:
> If you right-click on a pdf document attached to a page you should see an entry offering to copy a link to the document.
> Then you can go to any page in your wiki and insert the link (also edit the link text as usual).
> The link created should look like `[[../Task_3/Task_3-1/Test.pdf]]` in the page source.

Didn't get you. Maybe you misunderstood the goal here. I want to go to a page in the PDF, not to a page within the notebook.

**Comment by sanjarcode** [comment]
@pfalcon, I know what you are saying. It doesn't work because changing the default handling of files at the OS level will make it clumsy and frustrating to work with scripts attached to notebooks.

A similar alternative I had considered on Ubuntu is Nautilus(the file manager), which has a feature where you can set file handling behavior. It has 3 options:
1. Run the script
2. Display the file in the editor
3. Ask the user with a popup whether to run or display.

But `xdg-open`(which Zim uses internally) has nothing to do with Nautilus and and does not have the handling feature. Consequently, zim-wiki cannot use this popup feature. 

Now a plugin could surely force Zim to open attachements via Nautilus, but this approach is also bad:
1. It won't work on systems not having Nautilus.
2. Having a popup everytime you click a script is a pain if you are learning about scripts themselves.
3. You can't expect everyone to have the same MIME type behavior when the notebook is shared

**Comment by pfalcon** [comment]
xdg-open is configurable (to your likes), read its docs.

> clumsy and frustrating

Your alternative is to patch Zim with an adhoc feature. For that you need: a) to implement it at all; b) to refactor it to be implemented properly; c) get it accepted into the mainline (or it's just the same as you just configured your system work like you want in the first place - other users won't benefit from it automatically); d) once you switch away from Zim (very likely in the modern software landscape, new lures pop up every day), you're back in the cold with nothing working for you.

Now tell me how a), b), and c) compare in "clumsiness and frustration" with just getting habit to configure your system to work as *you* need. (That would take you a couple of years to collect reasonable stats to compare 2 approaches.)

**Comment by sanjarcode** [comment]
@pfalcon, an adhoc change is indeed undesirable. Using a plugin is the only solution here AFAIK. The plugin I mentioned in the previous message is actually the demonstration of a bad idea, because it blindly forces Zim to use Nautilus. A correct plugin would actually handle links only if href had some special suffix(like `#page`), and will stay silent if no such suffix exists. This way, there'd be a very small change to be made to Zim, but no change to `xdg-open` preferences. The behavior would be per notebook and contained within the plugin.

You are _correct_, If the plugin is disabled, Zim will have to atleast show 'plugin unavailable'.

**Comment by sanjarcode** [comment]
I was learning a new language and it occurred to me that 'small snippets of code' could be used to trigger for pronounciation. This is already used in proprietary notebook software like OneNote, the least we can do is have a plugin. I am busy currently, and would like to implement it myself.

**Comment by pfalcon** [comment]
> 'small snippets of code' could be used to trigger pronounciation

"Small snippets of code" could be used to ship a notebook which will call "rm -rf /" ("format c:" for Windows users among us) to your unsuspecting grandma. So, "small snippets of code" can never be a part of mainline, or even an "officially" listed plugin. (Individual users obviously may do whatever they want to their own Zim install or try to motivate other users to use their stuff).

**Comment by sanjarcode** [comment]
This is true about everything(in code), but thanks for pointing it out anyway.
"No code, no error" is not an option. And this is just a plugin.
My grandma doesn't care about zim-wiki, she'll probably use OneNote.

**Comment by pfalcon** [comment]
> "No code, no error(or catastrophe)" is so boring.

Well, that's not what I mean ;-). More like: "Someone who cares about *all* Zim users (up to wanting some feature be inside Zim, instead of letting individual users setup it outside of it), should also care about their/their data *security*".

> Now, my grandma wouldn't care about zim-wiki, she'll probably use OneNote.

That's pretty boring statement either ;-). I remember old good days when every open-source software strived for world domination ;-). All in all, it makes sense for Zim to be usable, reliable, and featureful for as wide array of users as possible, and 3rd goal is at odds with first 2.

**Comment by jaap-karssenberg** [comment]
Consider this a duplicate of #1412
